### PR TITLE
ROX-22718: Sort Roles in Tokens

### PR DESCRIPTION
**Pull request by aaa5kameric**
Sorting of roles

## Description

A detailed explanation of the changes in your PR.

Feel free to remove this section if it is overkill for your PR, and the title of your PR is sufficiently descriptive.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
Use this space to explain **how you validated** that **your change functions exactly how you expect it**.
Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by stehessel** [comment]
Hi @aaa5kameric,
thanks for your contribution, this issue has been bugging me for some time :). Could you please add a test for the service implementation that asserts the sorting in the list response? You can take inspiration from https://github.com/stackrox/stackrox/blob/a7ee107776bf374940c9dd7c1b691095caaa2b68/central/administration/events/service/service_impl_postgres_test.go to set up the test code.

File: central/apitoken/service/service_impl.go [hunk 1]
```diff
@@ -146,6 +147,7 @@ func (s *serviceImpl) ListAllowedTokenRoles(ctx context.Context, _ *v1.Empty) (*
 			result = append(result, role.GetRoleName())
 		}
 	}
+	sort.Strings(result)
```
Comment (RTann) [inline central/apitoken/service/service_impl.go:150]: https://pkg.go.dev/sort@go1.21.9#Strings is preferred for perf reasons (plus in 1.22, `sort.Strings` just calls `slices.Sort`)

```suggestion
	slices.Sort(result)
```

Comment (aaa5kameric) [reply central/apitoken/service/service_impl.go:150]: interesting, thank you for your comment. I see that sort.Strings(result) is a standard library function, while slices.Sort was also used in other scripts but as a custom defined function.  Such as the slices.Sort//Func sort for the missing tests, or the slices.SortFunc(oneOfFieldTypes, oneOfFieldTypeCmp). Thank you:)

File: central/apitoken/service/service_impl.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package service
 import (
 	"context"
 	"strings"
+	"sort"
```
Comment (RTann) [inline central/apitoken/service/service_impl.go:6]:
```suggestion
	"slices"
```
(plus it'd have to go above `"strings"`

File: central/apitoken/service/service_impl_test.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/auth/permissions"
+	"github.com/stackrox/rox/pkg/grpc/authn"
+	"github.com/stackrox/rox/pkg/grpc/authn/mocks"
+	"github.com/stackrox/rox/pkg/testutils/roletest"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/mock/gomock"
+)
+
+func TestSomethingIsSorted(t *testing.T) {
```
Comment (ludydoo) [inline central/apitoken/service/service_impl_test.go:19]: Would you mind renaming the test method?

**Comment by ludydoo** [comment]
pinging @stehessel

File: central/apitoken/service/service_impl_test.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	"go.uber.org/mock/gomock"
+)
+
+func TestServiceImpl_ListAllowedRoles_SortsRoleAlphabetically(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockIdentity := mocks.NewMockIdentity(mockCtrl)
+
+	mockDatastore := roleMock.NewMockDataStore(mockCtrl)
```
Comment (stehessel) [inline central/apitoken/service/service_impl_test.go:25]: I'd move this below to where it is first used.

File: central/apitoken/service/service_impl_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	roleOne := roletest.NewResolvedRole("Analyst", map[string]storage.Access{},
+		role.AccessScopeIncludeAll,
+	)
+
+	roleTwo := roletest.NewResolvedRole("Admin", map[string]storage.Access{},
+		role.AccessScopeIncludeAll,
+	)
+
+	roleThree := roletest.NewResolvedRole("Writer", map[string]storage.Access{},
```
Comment (stehessel) [inline central/apitoken/service/service_impl_test.go:34]: I would probably put `Writer` as the first role to make it more obvious that no sorting has been done yet.

File: central/apitoken/service/service_impl_test.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	mockDatastore.EXPECT().GetAllResolvedRoles(gomock.Any()).Return([]permissions.ResolvedRole{roleOne, roleTwo, roleThree}, nil)
+
+	s := &serviceImpl{roles: mockDatastore}
+
+	ctx := context.Background()
+	ctx = authn.ContextWithIdentity(ctx, mockIdentity, t)
+
+	actual, err := s.ListAllowedTokenRoles(ctx, &v1.Empty{})
+
+	assert.NoError(t, err)
```
Comment (stehessel) [inline central/apitoken/service/service_impl_test.go:49]:
```suggestion
	require.NoError(t, err)
```

File: central/apitoken/service/service_impl_test.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/stackrox/rox/pkg/testutils/roletest"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/mock/gomock"
+)
+
+func TestServiceImpl_ListAllowedRoles_SortsRoleAlphabetically(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockIdentity := mocks.NewMockIdentity(mockCtrl)
```
Comment (stehessel) [inline central/apitoken/service/service_impl_test.go:23]: same here

**Review by stehessel — COMMENTED** [review]
some minor comments

File: central/apitoken/service/service_impl_test.go (hunk 7 above)
Comment (aaa5kameric) [reply central/apitoken/service/service_impl_test.go:23]: Ok, thank you for your comments, I have implemented the requests and the test is still passed  👍

**Review by stehessel — APPROVED** [review]
LGTM, please format the files to fix the style linter error.

**Review by ludydoo — APPROVED** [review]
Same comment as @stehessel about linting

**Review by ludydoo — APPROVED** [review]

**Comment by ludydoo** [comment]
Once the checks are passing, you'll be able to hit that green "squash and merge" button 🥳

**Comment by ludydoo** [comment]
@stehessel do you know if it is expected that tests fails on external contributor forks? Looks like images are not pushed

**Comment by stehessel** [comment]
> @stehessel do you know if it is expected that tests fails on external contributor forks? Looks like images are not pushed

Yes believe that is intentional, see https://github.com/stackrox/stackrox/blob/master/.github/workflows/build.yaml#L388-L390.

**Comment by aaa5kameric** [comment]
Closing in favor of: https://github.com/stackrox/stackrox/pull/11614 because of CI failures

**Pull request closed without merging**